hydromt_wflow/workflows/river.py

```diff
@@ -291,6 +291,86 @@ def river_bathymetry(
     return ds_model[["rivwth", "rivdph"]]
 
 
+def river_floodplain_volume(
+    ds,
+    ds_model,
+    river_upa=30,
+    flood_depths=[0.5, 1.0, 1.5, 2.0, 2.5],
+    logger=logger,
+):
+    """Calculate the floodplain volume at given flood depths based on a (subgrid) HAND map.
+
+    Parameters
+    ----------
+    ds: xr.Dataset
+        hydrography dataset containing "flwdir", "uparea", "elevtn" variables;
+    ds_model: xr.Dataset, optional
+        Model dataset with output grid, must contain "rivmsk", "rivwth", "rivlen"
+        for subgrid must contain "x_out", "y_out".
+    river_upa: float
+        minimum threshold to define the river when calculating HAND, by default 30 [km2]
+    flood_depths : list of float, optional
+        flood depths at which a volume is derived, by default [0.5,1.0,1.5,2.0,2.5]
+    """
+    if not ds.raster.crs == ds_model.raster.crs:
+        raise ValueError("Hydrography dataset CRS does not match model CRS")
+
+    # check data variables.
+    dvars = ["flwdir", "elevtn", "uparea"]
+    dvars_model = ["rivmsk", "rivwth", "rivlen"]
+    if not np.all([v in ds for v in dvars]):
+        raise ValueError(f"One or more variables missing from ds: {dvars}.")
+    if not np.all([v in ds_model for v in dvars_model]):
+        raise ValueError(f"One or more variables missing from ds_model: {dvars_model}")
+
+    # initialize flood directions
+    flwdir = flw.flwdir_from_da(ds["flwdir"], mask=None)
+
+    # get river cell coordinates
+    riv_mask = ds_model["rivmsk"].values > 0
+    if "x_out" in ds_model and "y_out" in ds_model:  # use subgrid
+        idxs_out = ds.raster.xy_to_idx(
+            xs=ds_model["x_out"].values,
+            ys=ds_model["y_out"].values,
+            mask=riv_mask,
+            nodata=flwdir._mv,
+        )
+    else:
+        # get cell index of river cells
+        idxs_out = np.arange(ds_model.raster.size).reshape(ds_model.raster.shape)
+        idxs_out = np.where(riv_mask, idxs_out, flwdir._mv)
+
+    # calculate hand
+    hand = flwdir.hand(
+        drain=ds["uparea"].values >= river_upa, elevtn=ds["elevtn"].values
+    )
+
+    # calculate volume at user defined flood depths
+    # note that the ucat_map is not save currently but is required if
+    # we want to create a flood map by postprocessing
+    flood_depths = np.atleast_1d(flood_depths).ravel().astype(np.float32)  # force 1D
+    _, ucat_vol = flwdir.ucat_volume(idxs_out=idxs_out, hand=hand, depths=flood_depths)
+    # force minimum volume based on river width * length
+    min_vol = (
```

Comment:
It might also be useful to check the `min_vol` when the previous segment has a trapezoidal form? 

Then the `min_vol` used for comparison should also increase. Note, that this check is already done in Wflow.jl, and for the Meuse and the Rhine models the volume for quite some segments is adapted. For example, for the Meuse a total of 20128 rectangular segments are adapted for all river cells.

Reply:
Thanks for this comment @verseve! I've addressed this in the latest commit here. Can you check this.

I've tested it by checking it with the following code. However i still get a warning message when running the latest wflow.jl of the 1d_floodplain branch, which I think is a rounding error, but not sure how to best go about it. I'll add a comment about this over at https://github.com/Deltares/Wflow.jl/pull/233/
```
import xarray as xr
import numpy as np
ds = xr.open_dataset(r'staticmaps.nc')
da_fldvol = ds['floodplain_volume'].where(ds['wflow_river']==1, 0)
# second diff should be >= zero
check_vol = (da_fldvol.diff('flood_depth', n=2) >= 0)
assert np.all(check_vol) 
```

Reply:
I did check the changes in your latest commit, it looks good to me! I did add your suggested change for the warning message in Wflow.jl, I expect that with this change a warning message is not thrown. Is that the case?

Reply:
Thanks for the quick reply. I works as expected.